Content/Items/Magnet/Weapons.GrayGoo.cs

```diff
@@ -0,0 +1,435 @@
+			}
+		}
+
+		private void ReadjustDust()
+		{
+			NPC target = Main.npc[(int)EnemyWhoAmI];
+			foreach (Dust dust in Main.dust)
+			{
+				if (dust.type == ModContent.DustType<GrayGooDust>() && dust.customData is GrayGooDustData data && data.proj == Projectile)
+				{
+					Vector2 offset = Main.rand.NextVector2Circular(target.width / 2, target.height / 2);
+					data.x = (int)offset.X;
+					data.y = (int)offset.Y;
+				}
+			}
+		}
+
+		private void KillDust()
+		{
+			foreach (Dust dust in Main.dust)
+			{
+				if (dust.type == ModContent.DustType<GrayGooDust>() && dust.customData is GrayGooDustData data && data.proj == Projectile)
+				{
+					dust.active = false;
+				}
+			}
+		}
+
+		private static void DrawGooTarget(Projectile goo, SpriteBatch spriteBatch)
+		{
+			if (goo == default)
+				return;
+
+			var modproj = goo.ModProjectile as GrayGooProj;
+
+			if (!modproj.foundTarget)
+				return;
+
+			NPC NPC = Main.npc[(int)modproj.EnemyWhoAmI];
+
+			if (!NPC.active)
+				return;
+
+			if (NPC.active)
+			{
+				if (NPC.ModNPC != null)
+				{
+					if (NPC.ModNPC is ModNPC ModNPC)
+					{
+						if (ModNPC.PreDraw(spriteBatch, Main.screenPosition, NPC.GetAlpha(Color.White)))
+							Main.instance.DrawNPC((int)modproj.EnemyWhoAmI, false);
+
+						ModNPC.PostDraw(spriteBatch, Main.screenPosition, NPC.GetAlpha(Color.White));
+					}
+				}
+				else
+				{
+					Main.instance.DrawNPC((int)modproj.EnemyWhoAmI, false);
+				}
+			}
+		}
+		private void DrawNPCtarget(On.Terraria.Main.orig_CheckMonoliths orig)
+		{
+			orig();
+
+			var goos = Main.projectile.Where(n => n.active && n.type == ModContent.ProjectileType<GrayGooProj>()).ToList();
+			if (goos.Count == 0)
+				return;
+
+			GraphicsDevice gD = Main.graphics.GraphicsDevice;
+			SpriteBatch spriteBatch = Main.spriteBatch;
+
+			if (Main.gameMenu || Main.dedServ || spriteBatch is null || NPCTarget is null || gD is null)
+				return;
+
+			RenderTargetBinding[] bindings = gD.GetRenderTargets();
+			gD.SetRenderTarget(NPCTarget);
+			gD.Clear(Color.Transparent);
+
+			spriteBatch.Begin(default, default, default, default, default, null, Main.GameViewMatrix.ZoomMatrix);
+
+			goos.ForEach(n => DrawGooTarget(n, spriteBatch));
+			spriteBatch.End();
+			gD.SetRenderTargets(bindings);
+		}
+	}
+	public class GrayGooSplashDust : ModDust
+	{
+		public override string Texture => AssetDirectory.Invisible;
+
+		public override void OnSpawn(Dust dust)
+		{
+			dust.noGravity = false;
+			dust.noLight = false;
+		}
+
+		public override bool Update(Dust dust)
+		{
+			dust.position += dust.velocity;
+			dust.velocity.Y += 0.2f;
+			if (Main.tile[(int)dust.position.X / 16, (int)dust.position.Y / 16].HasTile && Main.tile[(int)dust.position.X / 16, (int)dust.position.Y / 16].BlockType == Terraria.ID.BlockType.Solid && Main.tileSolid[Main.tile[(int)dust.position.X / 16, (int)dust.position.Y / 16].TileType])
```

Comment:
This line is quite long, we should probably pull out local vars for these tiles. Additionally we should be using Framing.GetTileSafely instead of direct Main.tile accesses to prevent out-of-world access

Reply:
Fixed